runtime/callback.c

```diff
@@ -19,26 +19,51 @@
 
 #include <string.h>
 #include "caml/callback.h"
-#include "caml/domain.h"
+#include "caml/codefrag.h"
 #include "caml/fail.h"
+#include "caml/fiber.h"
 #include "caml/memory.h"
 #include "caml/mlvalues.h"
+#include "caml/platform.h"
+
+/*
+ * These macros are for ensuring effects are handled correctly
+ * inside callbacks. There are two aspects:
+ *  - we clear the stack parent for a callback to force an Unhandled
+ *  exception rather than effects being passed over the callback
+ *  - we register the stack parent as a local root while the callback
+ * is executing to ensure that the garbage collector follows the
+ * stack parent
+ */
+Caml_inline value save_and_clear_stack_parent(caml_domain_state* domain_state) {
+  struct stack_info* parent_stack = Stack_parent(domain_state->current_stack);
+  value cont = caml_alloc_1(Cont_tag, Val_ptr(parent_stack));
+  Stack_parent(domain_state->current_stack) = NULL;
+  return cont;
+}
+
+Caml_inline void restore_stack_parent(caml_domain_state* domain_state,
+                                      value cont) {
+  struct stack_info* parent_stack = Ptr_val(Op_val(cont)[0]);
+  CAMLassert(Stack_parent(domain_state->current_stack) == NULL);
+  Stack_parent(domain_state->current_stack) = parent_stack;
+}
+
+#define RESTORE_STACK_PARENT(domain_state) \
+
```

Comment:
This is a fragment left over from edits/merges, right?

```suggestion
```

Reply:
Fixed in 54365f2